### PR TITLE
ADX 984 implement o auth2 integration

### PR DESCRIPTION
**Pull request by kforenc**
## ADX 984 implement o auth2 integration

To allow access to ADR from 3rd party apps we use IAuthenticator plugin interface. After receiving request from another app, we analyze 'Authorization' HTTP Request header and check if it is a proper access_token. If yes we extract username from 'sub' claim (subject) and check it against database if user has already logged in. If not, we throw an exception.

Apart from that IMiddelware plugin interface is implemented to make one place for catching OAuth2-related own exceptions.

Moreover there was a need to customize WHO behaviour by implementing IChallengeDecider so that we don't get a redirect to login page after returning response with HTTP code 401.

relates [fjelltopp/adx_develop/141](https://github.com/fjelltopp/adx_develop/pull/141)

Tagged as v1.0.4

## Dependency Changes

New libraries added for access_token verification:
* python-dotenv
* python-jose

Setuptools was added in hard-coded version of 67.2.0 - any higher version was failing the build.

## Environment Changes

3 new variables added to adx_config.ini:
* ckanext.unaids.auth0_domain - auth0 domain that is used to verify incoming tokens
* ckanext.unaids.oauth2_api_audience - API Audience used by ADR API
* ckanext.unaids.oauth2_required_scope - hard-coded required scope

## Checklist

Put an `x` in the boxes that apply to this pull request (you can also fill these out after opening the pull request).
You may not need to check all boxes.

- [ ] The Jira ticket for this issue has been updated to "Ready to Review" or equivalent.
- [ ] I have developed these changes in discussion with the appropriate project manager.
- [ ] My code follows the general Fjelltopp documentation (see Confluence).
- [ ] I have made corresponding changes to the Fjelltopp documentation (see Confluence).
- [ ] I have rebased this branch with master.
- [ ] New dependency changes have been committed.
- [ ] I have added automated tests that prove my fix is effective or that my feature works.
- [ ] New and existing tests pass locally with my changes.
- [ ] My changes generate no new warnings.
- [ ] I have performed a self-review of my own code.
- [ ] I have assigned at least one reviewer.

### CODE REVIEW

**Review by tomeksabala — COMMENTED** [review]
One things worries me about this implementation:
You're using oauth2 tokens only as a wrapper and generate/use CKAN tokens in the background which are later used down the line.
1. You're using first of users api token which later might be misleading for the user. E.g. it might be a custom api token created by user for some API script and it will have incorrect "last time used" date
2. This solution might be obsolete later if we decide to use authorization scopes

I'd advise to look at other oauth2 implementation in ckan community. One example is this: https://github.com/datopian/ckanext-oauth2/blob/multi-sso-support/ckanext/oauth2/plugin.py where they overwrite IAuthenticator interface. Have you already considered this approach? What's your opinion on this?

File: ckanext/unaids/auth_logic.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+# based on work from https://auth0.com/docs/quickstart/backend/python/01-authorization
+def access_token_present_and_valid_and_user_authorized():
+    token = request.headers.get('Authorization', '')
+    if token:
+        access_token = validate_and_decode_token(token)
+        verify_required_scope(access_token)
+        subject = access_token['sub']
+
+        if not subject.startswith("auth0|"):
```
Comment (kforenc) [inline ckanext/unaids/auth_logic.py:51]: If we want to allow other identity providers (e.g. 'google-oauth2') we can change it into a variable. Or remove this check at all.

**Comment by kforenc** [comment]
I didn't do any E2E tests, but maybe we could do it using [this approach](https://auth0.com/docs/get-started/authentication-and-authorization-flow/call-your-api-using-resource-owner-password-flow) to get access_token for some user created strictly for testing.

**Review by tomeksabala — APPROVED** [review]
I have run pytest for this branch locally and got it passed. In my opinion this work is ready to be merged.
Can you squash merge it to the development branch, and then tag it with a new minor version X.Y+1.Z? After that's done please update the submodule/ckanext-unaids dependency in the main PR in adx_develop.

**Pull request merged**